### PR TITLE
Zz/merge upstream 0902

### PR DESCRIPTION
**Pull request by zhenfeizhang**

### CODE REVIEW

**Comment by zhenfeizhang** [comment]
This PR is ready for review. Though it may be better to create a separate branch for it (i.e., `scroll-dev-0902`) than merging it back to `scroll-dev-0607`

**Comment by lispc** [comment]
@zhenfeizhang the git history is a bit strange. I think we would have a PR with 10 commits rather than 200 commits?

File: halo2_proofs/src/dev.rs [hunk 1]
```diff
@@ -565,354 +564,7 @@ impl<F: FieldExt> MockProver<F> {
...
-            errors.dedup_by(|a, b| match (a, b) {
-                (
-                    a @ VerifyFailure::ConstraintPoisoned { .. },
-                    b @ VerifyFailure::ConstraintPoisoned { .. },
-                ) => a == b,
-                _ => false,
-            });
-            Err(errors)
-        }
+        self.verify_at_rows_par(self.usable_rows.clone(), self.usable_rows.clone())
```
Comment (lispc) [inline halo2_proofs/src/dev.rs:567]: it works. But sometimes we deliberately use single thread in proving. (eg: cargo test launch tests parallelly so it bring no benefit (if not worse) to use multi thread innerly.

**Comment by lispc** [comment]
[4f34026](https://github.com/scroll-tech/halo2/pull/8/commits/4f340263e4f2d7c97a9ea2492726d1b6b2c75fc7)
[502d7dd](https://github.com/scroll-tech/halo2/pull/8/commits/502d7dd32d35cc665674749039b71b17bbaf369d)
[7479c9e](https://github.com/scroll-tech/halo2/pull/8/commits/7479c9ebdbb9bb5d01a3ec1b564f653e65552492)

are useful for our prover. Seems not here yet?

File: halo2_proofs/src/dev.rs (hunk 1 above)
Comment (zhenfeizhang) [reply halo2_proofs/src/dev.rs:567]: right. Seems like in the PSE's integration, they have removed `self.verify_at_rows(...)` and the only option here is `self.verify_at_rows_par`

**Comment by zhenfeizhang** [comment]
> [4f34026](https://github.com/scroll-tech/halo2/pull/8/commits/4f340263e4f2d7c97a9ea2492726d1b6b2c75fc7) [502d7dd](https://github.com/scroll-tech/halo2/pull/8/commits/502d7dd32d35cc665674749039b71b17bbaf369d) [7479c9e](https://github.com/scroll-tech/halo2/pull/8/commits/7479c9ebdbb9bb5d01a3ec1b564f653e65552492)
> 
> are useful for our prover. Seems not here yet?

Fixed the first and third ones with 4b53eee6d3f2db2f525e69d2f0d8baddeb7e417c.

IIUC, the second one was integrated by PSE already:

``` rust
 inputs
      .par_iter()
      .filter_map(move |(input, input_row)| {
          if table.binary_search(input).is_err() {
              Some(VerifyFailure::Lookup {
                  name: lookup.name,
                  lookup_index,
                  location: FailureLocation::find_expressions(
                      &self.cs,
                      &self.regions,
                      *input_row,
                      lookup.input_expressions.iter(),
                  ),
              })
          } else {
              None
          }
      })
      .collect::<Vec<_>>()
```

**Comment by lispc** [comment]
close for https://github.com/scroll-tech/halo2/pull/15

**Pull request closed without merging**